File: invertBinaryTree/invertBinaryTree.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution{
public:
  TreeNode* invertTree(TreeNode *root) {
    /* if (!root)
      return NULL;
    TreeNode *t = new TreeNode(root->val);
    t->left = invertTree(root->right);
    t->right = invertTree(root->left);
    return t; */
    if (!root)
      return NULL;
    TreeNode *temp = invertTree(root->left);
    root->left = invertTree(root->right);
    root->right = temp;
    return root;
  }
};

int main()
{
  Solution s;
  int tree[7] = {4,2,7,1,3,6,9};
  int size = sizeof(tree)/sizeof(tree[0]);
  vector<TreeNode *> vec;
  for (int i = 0; i < size; i++) {
    if (tree[i] != 9999) {
      vec.push_back(new TreeNode(tree[i]));
    } else {
      vec.push_back(NULL);
    }
  }
  for (int i = 0; i < size/2; i++) {
    if (i*2 + 1 < size)
      vec[i]->left = vec[i*2 + 1];
    if (i*2 + 2 < size)
      vec[i]->right = vec[i*2 + 2];
  }
  TreeNode *root = s.invertTree(vec[0]);
  queue<TreeNode*> q;
  q.push(root);
  while (!q.empty()) {
    TreeNode *t = q.front();
    cout << t->val << " ";
    q.pop();
    if (t->left)
      q.push(t->left);
    if (t->right)
      q.push(t->right);
    //delete t; 
  }
  cout << endl;
  for (TreeNode *t : vec) {
    delete t;
  }
  vec.clear();
  return 0;
}

File: 226.invertBinaryTree/invertBinaryTree.java
// javac InvertBinaryTree.java
// java InvertBinaryTree
import java.util.*;

class TreeNode {
	int val;
	TreeNode left;
	TreeNode right;
	TreeNode(int x) { val = x; }
}

public class InvertBinaryTree {
    public TreeNode invertTree(TreeNode root) {
        if (root == null)
            return root;
        TreeNode left = invertTree(root.left);
        TreeNode right = invertTree(root.right);
        root.left = right;
        root.right = left;
        return root;
    }
	public static void main(String[] args) {
		int tree[] = {4,2,7,1,3,6,9};
		ArrayList<TreeNode> arrayList = new ArrayList<>();
		for (int i = 0; i < tree.length; i++) {
			if (tree[i] != 9999)
				arrayList.add(new TreeNode(tree[i]));
			else
				arrayList.add(null);
		}
		for (int i = 0; i < tree.length / 2; i++) {
			if (arrayList.get(i) == null)
				continue;
			if (i * 2 + 1 < tree.length)
				arrayList.get(i).left = arrayList.get(i * 2 + 1);
			if (i * 2 + 2 < tree.length)
				arrayList.get(i).right = arrayList.get(i * 2 + 2);
		}
		InvertBinaryTree s = new InvertBinaryTree();
		TreeNode root = s.invertTree(arrayList.get(0));
		Queue<TreeNode> q = new LinkedList<TreeNode>();
		q.offer(root);
		while (q.size() != 0) {
			TreeNode t = q.poll();
			System.out.print(t.val + " ");
			if (t.left != null)
				q.offer(t.left);
			if (t.right != null)
				q.offer(t.right);
		}
		System.out.println();
	}
}
